ui/src/app/components/breadcrumb.tsx

```diff
@@ -27,75 +34,99 @@ const styles = {
+        new BreadcrumbData('Notebooks', `${prefix}/notebooks`),
+        new BreadcrumbData(
+          nbName && dropNotebookFileSuffix(decodeURIComponent(nbName)),
+          `${prefix}/notebooks/${nbName}`
+        )
       ];
     case BreadcrumbType.ConceptSet:
       return [
-        ...getTrail(BreadcrumbType.Data, data),
-        {label: conceptSet ? conceptSet.name : '...', url: `${prefix}/data/concepts/sets/${csid}`}
+        ...getTrail(BreadcrumbType.Data, workspace, cohort, conceptSet, urlParams),
+        new BreadcrumbData(
+          conceptSet
+            ? conceptSet.name
+            : '...', `${prefix}/data/concepts/sets/${csid}`
+        )
       ];
     case BreadcrumbType.Cohort:
       return [
-        ...getTrail(BreadcrumbType.Data, data),
-        {label: cohort ? cohort.name : '...',
-          url: `${prefix}/data/cohorts/${cid}/review/participants`}
+        ...getTrail(BreadcrumbType.Data, workspace, cohort, conceptSet, urlParams),
+        new BreadcrumbData(
+          cohort ? cohort.name : '...',
+          `${prefix}/data/cohorts/${cid}/review/participants`
+        )
       ];
     case BreadcrumbType.Participant:
       return [
-        ...getTrail(BreadcrumbType.Cohort, data),
-        {label: `Participant ${pid}`,
-          url: `${prefix}/data/cohorts/${cid}/review/participants/${pid}`}
+        ...getTrail(BreadcrumbType.Cohort, workspace, cohort, conceptSet, urlParams),
+        new BreadcrumbData(
+          `Participant ${pid}`,
+          `${prefix}/data/cohorts/${cid}/review/participants/${pid}`
+        )
       ];
     case BreadcrumbType.CohortAdd:
       return [
-        ...getTrail(BreadcrumbType.Data, data),
-        {label: 'Build Cohort Criteria', url: `${prefix}/data/cohorts/build`}
+        ...getTrail(BreadcrumbType.Data, workspace, cohort, conceptSet, urlParams),
+        new BreadcrumbData('Build Cohort Criteria', `${prefix}/data/cohorts/build`)
       ];
     case BreadcrumbType.SearchConcepts:
       return [
-        ...getTrail(BreadcrumbType.Data, data),
-        {label: 'Search Concepts', url: `${prefix}/data/concepts`}
+        ...getTrail(BreadcrumbType.Data, workspace, cohort, conceptSet, urlParams),
+        new BreadcrumbData('Search Concepts', `${prefix}/data/concepts`)
       ];
     case BreadcrumbType.Dataset:
       return [
-        ...getTrail(BreadcrumbType.Data, data),
-        {label: 'Dataset', url: `${prefix}/data/datasets`}
+        ...getTrail(BreadcrumbType.Data, workspace, cohort, conceptSet, urlParams),
+        new BreadcrumbData('Dataset', `${prefix}/data/datasets`)
       ];
     case BreadcrumbType.Data:
       return [
-        ...getTrail(BreadcrumbType.Workspaces, data),
-        {label: workspace ? workspace.name : '...', url: `${prefix}/data`}
+        ...getTrail(BreadcrumbType.Workspaces, workspace, cohort, conceptSet, urlParams),
+        new BreadcrumbData(workspace ? workspace.name : '...', `${prefix}/data`)
```

Comment:
How about creating a map with key as BreadCrumbType and values {label, url} that ways this switch case, can be easily replaced by .
..new BreadCrumbData(mapname.get(BreadCrumbtype).label,mapname.get(BreadCrumbtype).url)
Also that will help in case there is a new breadcrumbType too

Reply:
Because the labels / urls aren't 1:1 with only the `BreadcrumbType`, this map would end up doing basically the same thing as the switch statement.